Overlay options on restart from checkpoint

fmahebert opened the pull request:
## Proposed changes

Adds support for updating simulation parameters at restarts from checkpoint.

The interface is fairly rigid, but should be sufficient for a first pass:
- Looks for file `OriginalInputFile.overlay000123.yaml` when restarting from checkpoint `000123`.
- Allows certain GlobalCache tags only to be updated

Things I'm not sure about:
- The names of everything. The word "overlay" makes good sense at the file parser level, but not at the user-level when describing the feature "let's update parameters when restarting from checkpoint". I picked some names for the phase and functions but probably they can be improved on.
- The way Charm++ restarts work, the code doesn't know it's restarting from checkpoint vs just continuing as part of a longer run. In other words the phase sequence "WriteCheckpoint -> UpdateOptions -> Evolve" looks the same whether it's happening 10 hours into a run or 0 hours into a run. Right now I've added the reparsing phase into the wallclock-aware phase control, which lets us reparse only when restarting. But if someone writes `VisitAndReturn(WriteCheckpoint), VisitAndReturn(UpdateOptions)` phase control, the code might try to reparse input file every N slabs for example. I don't love it, but maybe this is just how the feature works...

The code isn't totally done: some tests still need to be written. I'm putting in this draft PR so we can start discussing the design, and because I will likely not have time to write the tests in the next few days anyway... In this draft the tests will fail because I've tweaked the Burgers input files to demonstrate how the feature works, but this is not currently in a "good state".

To try the feature yourself:
- compile `EvolveBurgersStep`
- `bin/EvolveBurgersStep ../tests/InputFiles/Burgers/Step.yaml`<- runs for ~1 minute, checkpoints and exits
- `bin/EvolveBurgersStep +restart SpectreCheckpoint000000` <- no overlay file, proceeds without changes
- `bin/EvolveBurgersStep +restart SpectreCheckpoint000001` <- overlay file changes obs frequency
- `bin/EvolveBurgersStep +restart SpectreCheckpoint000002` <- overlay file changes obs frequency again
- continue ad nauseum <- no more overlay files, proceeds without changes

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

fmahebert committed:
Promote Main's Option::Parser to a member variable

fmahebert committed:
Fix location of Doxygen group close

fmahebert committed:
Promote Main's input_file to a member variable

fmahebert committed:
Enable overlaying the EventsAndTriggers option

fmahebert committed:
Add helper to identify tags that can be updated on restart

fmahebert committed:
Add GlobalCache interface for overlaying new values

fmahebert committed:
Add a new phase UpdateOptionsAtRestartFromCheckpoint

fmahebert committed:
Add update options phase to checkpoint phase control

fmahebert committed:
Add update options phase to evolution executables

fmahebert committed:
TESTING Add overlay input files for Burgers Step problem

fmahebert committed:
Discuss overlaying in checkpoint documentation

wthrowe commented on src/Parallel/ParallelComponentHelpers.hpp:
```diff
@@ -293,6 +294,51 @@ using get_option_tags = tmpl::remove_duplicates<tmpl::flatten<
+template <typename Tag>
+struct is_tag_overlayable<Tag, std::void_t<decltype(Tag::is_overlayable)>>
+    : std::bool_constant<Tag::is_overlayable> {
+  static_assert(tmpl::size<typename Tag::option_tags>::value == 1,
```
Include `not is_overlayable or` in this.  Although it looks like you have a similar assertion where this restriction is actually assumed, so maybe just remove this one and rely on the check there.

wthrowe commented on docs/Tutorials/CheckpointRestart.md:
```diff
@@ -53,3 +51,39 @@ and restarting:
+`CheckpointAndExitAfterWallclock` phase control automatically directs code flow
+to this phase after a restart.
+
+TODO FOR CODE REVIEW: If someone uses a
+`VisitAndReturn(UpdateOptionsAtRestartFromCheckpoint)` just after a
+`VisitAndReturn(WriteCheckpoint)`, the code will try to parse the overlay files
+during evolution, even if no restarts occur. In other words, the code might
+write a checkpoint every 100 slabs, and will try to reparse the input file after
+each one. Since it's hard to know if there was a restart in your past (except
+perhaps by making guesses from the wallclock time?), it's hard to handle this
+sanely...
```
If Main cannot migrate (or if you can distinguish a migration from a checkpoint), you could keep track of whether a restart has completed using a non-checkpointed member variable.

wthrowe reviewed (commented):
I think this is a reasonable design.  I'm not sure what a good name would be for the files.  Overlay, updates, changes.

fmahebert committed:
Fixup: promote parser to member variable

fmahebert committed:
Fixup: remove a redundant static assert

fmahebert committed:
Fixup: Make sure options are only updated in first phase after restart from checkpoint

fmahebert commented:
Rebased, and added some fixups for @wthrowe's review and to fix a comment I'd neglected to update.

@wthrowe: adding an uncheckpointed bool to Main was a good suggestion! I've implemented and checked this works as expected: if I now use `VisitAndReturn(WriteCheckpoint)` and `VisitAndReturn(UpdateOptions...)` together, the options are only updated at the time of restart.

knelli2 commented:
See #5585 for updated version of this PR
